File: src/fs_explorer/document_pages.py
"""Helpers for page manifests backed by object-store markdown files."""

from __future__ import annotations

from pathlib import Path
from typing import Any

from .blob_store import BlobStore, LocalBlobStore, resolve_object_store_dir
from .page_store import parse_page_front_matter
from .storage import DocumentPageRecord, PostgresStorage


def page_record_from_manifest(item: Any, *, document_id: str | None = None) -> DocumentPageRecord:
    """Coerce one manifest row into the storage dataclass."""
    resolved_document_id = document_id or (
        getattr(item, "document_id", None)
        if not isinstance(item, dict)
        else item.get("document_id")
    )
    if resolved_document_id is None:
        raise ValueError("document_id is required to build a page record.")
    return DocumentPageRecord(
        document_id=str(resolved_document_id),
        page_no=int(item.page_no if hasattr(item, "page_no") else item["page_no"]),
        object_key=str(item.object_key if hasattr(item, "object_key") else item["object_key"]),
        heading=(
            item.heading if hasattr(item, "heading") else item.get("heading")
        ),
        source_locator=(
            item.source_locator
            if hasattr(item, "source_locator")
            else item.get("source_locator")
        ),
        content_hash=str(
            item.content_hash if hasattr(item, "content_hash") else item["content_hash"]
        ),
        char_count=int(item.char_count if hasattr(item, "char_count") else item["char_count"]),
        is_synthetic_page=bool(
            item.is_synthetic_page
            if hasattr(item, "is_synthetic_page")
            else item.get("is_synthetic_page", False)
        ),
    )


def resolve_pages_directory(
    *,
    blob_store: BlobStore,
    pages_prefix: str,
) -> str:
    """Return the local directory path for one document pages prefix."""
    prefix = str(pages_prefix or "").strip().rstrip("/")
    if not prefix:
        return ""
    if isinstance(blob_store, LocalBlobStore):
        return str((blob_store.root_dir / prefix).resolve())
    return str((resolve_object_store_dir() / prefix).resolve())


def read_page_content(
    *,
    blob_store: BlobStore,
    page: dict[str, Any],
) -> dict[str, Any]:
    """Load one page blob and merge front matter with manifest metadata."""
    raw = blob_store.get(object_key=str(page["object_key"])).decode("utf-8")
    header, body = parse_page_front_matter(raw)
    return {
        "page_no": int(page["page_no"]),
        "unit_no": int(page["page_no"]),
        "object_key": str(page["object_key"]),
        "heading": page.get("heading") or header.get("heading") or None,
        "source_locator": page.get("source_locator") or header.get("source_locator") or None,
        "content_hash": str(page["content_hash"]),
        "char_count": int(page.get("char_count") or len(body)),
        "is_synthetic_page": bool(page.get("is_synthetic_page", False)),
        "page_label": header.get("page_label") or str(page["page_no"]),
        "markdown": body,
        "file_path": blob_store.materialize(object_key=str(page["object_key"])),
    }


def load_document_pages(
    *,
    storage: PostgresStorage,
    blob_store: BlobStore,
    document_id: str,
    page_nos: list[int] | None = None,
) -> list[dict[str, Any]]:
    """Load manifest rows plus markdown content for one document."""
    pages = storage.list_document_pages(document_id=document_id, page_nos=page_nos)
    return [read_page_content(blob_store=blob_store, page=page) for page in pages]


def find_page_by_path(
    *,
    storage: PostgresStorage,
    blob_store: BlobStore,
    document_ids: list[str],
    file_path: str,
) -> tuple[dict[str, Any], dict[str, Any]] | None:
    """Resolve one page file path back to (document, page)."""
    normalized = str(Path(file_path).resolve()).lower()
    documents = storage.list_documents_by_ids(doc_ids=document_ids, include_deleted=False)
    for document in documents:
        pages = storage.list_document_pages(document_id=str(document["id"]))
        for page in pages:
            local_path = blob_store.materialize(object_key=str(page["object_key"]))
            if str(Path(local_path).resolve()).lower() == normalized:
                return document, read_page_content(blob_store=blob_store, page=page)
    return None

File: src/runtime/document-pages.ts
/*
Reference: legacy/python/src/fs_explorer/document_pages.py
Reference: legacy/python/src/fs_explorer/page_store.py
*/

import { resolve } from "node:path";

import { LocalBlobStore, resolveObjectStoreDir } from "./blob-store.js";
import { parsePageFrontMatter } from "./page-store.js";
import type { BlobStore, LoadedDocumentPage, StoredPage } from "../types/library.js";
import type {
  SqliteStorageBackend,
  StorageDocumentPageRecord,
  StoredDocument,
  StoredDocumentPage,
} from "../types/storage.js";

export function pageRecordFromManifest(
  item: StoredPage | StoredDocumentPage,
  input: { documentId?: string | null } = {},
): StorageDocumentPageRecord {
  const documentId = input.documentId ?? ("document_id" in item ? item.document_id : null);
  if (!documentId) {
    throw new Error("document_id is required to build a page record.");
  }
  return {
    documentId,
    pageNo: "pageNo" in item ? item.pageNo : item.page_no,
    objectKey: "objectKey" in item ? item.objectKey : item.object_key,
    heading: item.heading ?? null,
    sourceLocator: "sourceLocator" in item ? item.sourceLocator ?? null : item.source_locator ?? null,
    contentHash: "contentHash" in item ? item.contentHash : item.content_hash,
    charCount: "charCount" in item ? item.charCount : item.char_count,
    isSyntheticPage:
      "isSyntheticPage" in item ? item.isSyntheticPage : item.is_synthetic_page,
  };
}

export function resolvePagesDirectory(input: {
  blobStore: BlobStore;
  pagesPrefix: string;
}): string {
  const prefix = String(input.pagesPrefix || "").trim().replace(/\/+$/, "");
  if (!prefix) {
    return "";
  }
  if (input.blobStore instanceof LocalBlobStore) {
    return resolve(input.blobStore.rootDir, prefix);
  }
  return resolve(resolveObjectStoreDir(), prefix);
}

export async function readPageContent(input: {
  blobStore: BlobStore;
  page: StoredDocumentPage;
}): Promise<LoadedDocumentPage> {
  const raw = Buffer.from(await input.blobStore.get({ objectKey: input.page.object_key })).toString("utf8");
  const [header, body] = parsePageFrontMatter(raw);
  return {
    unit_no: input.page.page_no,
    page_no: input.page.page_no,
    object_key: input.page.object_key,
    heading: input.page.heading ?? header.heading ?? null,
    source_locator: input.page.source_locator ?? header.source_locator ?? null,
    content_hash: input.page.content_hash,
    char_count: input.page.char_count || body.length,
    is_synthetic_page: input.page.is_synthetic_page,
    page_label: header.page_label ?? String(input.page.page_no),
    markdown: body,
    file_path: await input.blobStore.materialize({ objectKey: input.page.object_key }),
  };
}

export async function loadDocumentPages(input: {
  storage: SqliteStorageBackend;
  blobStore: BlobStore;
  documentId: string;
  pageNos?: number[] | null;
}): Promise<LoadedDocumentPage[]> {
  const pages = input.storage.listDocumentPages(input.documentId, input.pageNos ?? null);
  const loaded: LoadedDocumentPage[] = [];
  for (const page of pages) {
    loaded.push(await readPageContent({ blobStore: input.blobStore, page }));
  }
  return loaded;
}

export async function findPageByPath(input: {
  storage: SqliteStorageBackend;
  blobStore: BlobStore;
  documentIds: string[];
  filePath: string;
}): Promise<{ document: StoredDocument; page: LoadedDocumentPage } | null> {
  const normalized = resolve(input.filePath).toLowerCase();
  const documents = input.storage.listDocumentsByIds(input.documentIds, false);
  for (const document of documents) {
    const pages = input.storage.listDocumentPages(document.id);
    for (const page of pages) {
      const localPath = await input.blobStore.materialize({ objectKey: page.object_key });
      if (resolve(localPath).toLowerCase() === normalized) {
        return {
          document,
          page: await readPageContent({ blobStore: input.blobStore, page }),
        };
      }
    }
  }
  return null;
}
